core/src/main/scala/cats/data/HashSet.scala

```diff
@@ -0,0 +1,937 @@
+      case _ =>
+        false
+    }
+
+    override def toString(): String = {
+      val valueMapStr =
+        ("0" * Integer.numberOfLeadingZeros(if (valueMap != 0) valueMap else 1)) + Integer.toBinaryString(valueMap)
+      val nodeMapStr =
+        ("0" * Integer.numberOfLeadingZeros(if (nodeMap != 0) nodeMap else 1)) + Integer.toBinaryString(nodeMap)
+
+      s"""BitMapNode(valueMap=$valueMapStr, nodeMap=$nodeMapStr, contents=${contents.mkString("[", ", ", "]")})"""
+    }
+  }
+
+  object Node {
+    val BitPartitionSize = 5
+    val BitPartitionMask = (1 << BitPartitionSize) - 1
+    val MaxDepth = 7
+
+    val SizeNone = 0
+    val SizeOne = 1
+    val SizeMany = 2
+
+    /**
+      * The `mask` is a 5-bit segment of a 32-bit element hash.
+      *
+      * The `depth` value is used to determine which segment of the hash we are currently inspecting by shifting the `elementHash` to the right in [[Node.BitPartitionSize]] bit increments.
+      *
+      * A 5-bit segment of the hash can represent numbers 0 to 31, which matches the branching factor of the trie structure.
+      *
+      * It represents the notional index of the element in the current trie node.
+      *
+      * @param elementHash the hash of the element we are operating on.
+      * @param depth the depth of the current node in the trie structure.
+      * @return the relevant 5-bit segment of the `elementHash`.
+      */
+    def maskFrom(elementHash: Int, depth: Int): Int =
+      (elementHash >>> (depth * Node.BitPartitionSize)) & BitPartitionMask
+
+    /**
+      * Sets a single bit at the position of the notional index indicated by `mask`.
+      *
+      * Used to determine the bit which represents the notional index of a data value or node in the trie node bitmaps.
+      *
+      * @param mask the notional index of an element at this depth in the trie.
+      * @return an integer with a single bit set at the notional index indicated by `mask`.
+      */
+    def bitPosFrom(mask: Int): Int =
+      1 << mask
+
+    /**
+      * Calculates the absolute index of an element in the contents array of a trie node.
+      *
+      * This is calculated by counting how many bits are set to the right of the notional index in the relevant bitmap.
+      *
+      * @param bitMap the bitmap indicating either data value or node positions in the contents array.
+      * @param bitPos the notional index of the element in the trie node.
+      * @return the absolute index of an element in the contents array.
+      */
+    def indexFrom(bitMap: Int, bitPos: Int): Int =
+      Integer.bitCount(bitMap & (bitPos - 1))
+
+    /**
+      * Creates a new empty bitmap node.
+      *
+      * @param hash the [[cats.kernel.Hash]] instance to use to hash elements.
+      * @return a new empty bitmap node.
+      */
+    def empty[A](implicit hash: Hash[A]): Node[A] =
+      new BitMapNode(0, 0, Array.empty[Any], 0)
+  }
+
+  private[data] class Iterator[A] extends scala.Iterator[A] {
```

Comment:
It probably doesn't matter much, but there is a `scala.collection.AbstractIterator`, which is an abstract class (instead of a trait). (Also for the reverse iterator below.)

Reply:
Ahhhh thanks that's cool, should reduce bytecode size